Cookbook/Proposals/SwiftGen.md

```diff
@@ -0,0 +1,83 @@
+## Adopt SwiftGen for localizable strings
+
+* Author: David Rodrigues
+* Review Manager: Danilo Aliberti
+
+### Introduction
+
+Localisable strings by themselves are not a hard thing to use but for apps with a significant number of strings it's entirely impossible to remember or discover easily the respective key to use.
+
+```swift
+title = NSLocalizedString("hello_world_title", comment: "")
+```
+
+Because of this, and considering the size of our codebase, we use concrete types that map our localisable strings which enables the usage of autocomplete.
+
+```swift
+public struct LocalizationUI {
+    public struct Common {
+        public static var ok: String { return localizedName(key: "general_ok") }
+        ...
+    }
+}
+```
+
+### Problems
+
+The current approach works, as we would expect, but there's a couple of caveats in it.
+
+##### 1. Maintenance
+
+This types at the moment are created and managed by us manually which means any new addition/edition/removal needs to be done by us. This may not look very significant but considering the number of strings, more than 3k, it is a lot to be maintained/extended.
+
+##### 2. Safety
+
+While this types are useful for development by enabling autocomplete they don't provide any safety since we can add a new property for a new key without necessarily adding the new string.
+
+### Proposed Solution
+
+SwiftGen is a code generator which is able to generate all this types and properties automatically by running a script which would fix the two main problems that we have with the current approach.
+
+- Maintenance will be zero because everything will be mapped based on the localisable strings;
+- Compile-time type safety will be guaranteed because everything will be generated from the localizable strings making it impossible to use something without being in the strings file or keep using something that was deleted.
```

Comment:
This comes with a tradeoff that if we need specific string with a different key for one particular target we will need to put it into common strings file. I guess using SwiftGen will affect how we approach this and target specific strings as well, because i.e. code generated for target specific strings may collide with common generated code.

Reply:
I think this won't be a problem @ilyapuchka because we will need our own function to obtain the strings in order to keep the existent flow.

1. Check target specific table
2. If available return
3. If not check the main table and return

Reply:
That's in fact something that I've done in past projects, with a simple change via a custom template, so it's definitively doable.